### PR TITLE
Salto 1568: salesforce inconsistent fetch results

### PR DESCRIPTION
**Pull request by tamtamirr**
The filter `addMissingIds` in Salesforce used to cause inconsistent fetches since it would shut down immediately when the operation failed on a specific type.

---

Using `Promise.allSettled` fixes the problem.

---
_Release Notes_: 
Monorepo:
- Upgraded ECMAScript version to `es2020`

Salesforce:
- `addMissingIds` filter will not result in inconsistent fetches.

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/add_missing_ids.ts [hunk 1]
```diff
@@ -74,22 +69,23 @@ export const WARNING_MESSAGE = 'Encountered an error while trying populate inter
...
-        metadataType,
-      )
-      log.debug(`Getting missing ids for the following types: ${Object.keys(groupedElements)}`)
-      await Promise.all(
-        Object.entries(groupedElements)
-          .map(([typeName, typeElements]) => addMissingIds(client, typeName, typeElements))
-      )
-    },
-  }),
+  onFetch: async (elements: Element[]) => {
```
Comment (netama) [inline packages/salesforce-adapter/src/filters/add_missing_ids.ts:72]: why did you revert the `ensureSafeFilterFetch`? why not keep it and just change the `allSettled`? (also notice the next line isn't doing anything - but either way I think we shouldn't disable it, at least until we see this working consistently)

File: tsconfig.base.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
         "declarationMap": true,
         "module": "commonjs",
         "esModuleInterop": true,
-        "target": "es2019",
+        "target": "es2020",
```
Comment (netama) [inline tsconfig.base.json:7]: @ori-moisis anything else that should be checked here?

File: packages/salesforce-adapter/src/filters/add_missing_ids.ts (hunk 1 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/filters/add_missing_ids.ts:72]: we should definitely not remove the `ensureSafeFilterFetch`

Comment (tamtamirr) [reply packages/salesforce-adapter/src/filters/add_missing_ids.ts]: v, reverted

File: packages/salesforce-adapter/test/filters/add_missing_ids.test.ts [hunk 3]
```diff
@@ -195,26 +199,8 @@ describe('Internal IDs filter', () => {
     })
   })
 
-  describe('when feature is throwing an error', () => {
```
Comment (ori-moisis) [inline packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:198]: why?

File: packages/salesforce-adapter/test/filters/add_missing_ids.test.ts [hunk 4]
```diff
@@ -13,14 +13,18 @@
...
+import filterCreator from '../../src/filters/add_missing_ids'
 import mockClient from '../client'
 import {
-  SALESFORCE, API_NAME, METADATA_TYPE, INSTANCE_FULL_NAME_FIELD, INTERNAL_ID_ANNOTATION,
+  API_NAME,
+  INSTANCE_FULL_NAME_FIELD,
+  INTERNAL_ID_ANNOTATION,
   INTERNAL_ID_FIELD,
+  METADATA_TYPE,
+  SALESFORCE,
```
Comment (ori-moisis) [inline packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:27]: why?

Comment (tamtamirr) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:27]: lint fix

File: packages/salesforce-adapter/test/filters/add_missing_ids.test.ts (hunk 3 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:198]: Promise.allSettled doesnt throw an error

Comment (ori-moisis) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:198]: ok, so the test found that you broke the expected behavior
you should fix it, not remove the test

Comment (tamtamirr) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:198]: added a warning message when one of the promises was rejected + test

Comment (ori-moisis) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:198]: you still changed the behavior in a way the breaks the previous interface
what is the reason to change the behavior as part of this fix?

Comment (tamtamirr) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:198]: The "feature" throwing the error wasn't the bug we were fixing in this PR? Promise.all rejected one of the types and threw an error, Promise.allSettled executes everything and returns report on each promise. This method doesn't throw anymore (That's why in one of the previous CR commits I removed `ensureSafeFilterFetch`). I'm a little bit confused and may not understand this, we can discuss F2F if needed :)

Comment (ori-moisis) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts:198]: the issue was that it returned a result before it actually finished running things.
the next filter would start running while this filter would still have code executing which is not good.

if something fails we still want this to fail. we don't want to change any behavior, we only have a problem with the order of execution

Comment (tamtamirr) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts]: v, reverted changes in tests

File: packages/salesforce-adapter/test/filters/add_missing_ids.test.ts (hunk 4 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/test/filters/add_missing_ids.test.ts]: reverted

File: packages/salesforce-adapter/src/filters/add_missing_ids.ts [hunk 5]
```diff
@@ -84,10 +79,13 @@ const filter: FilterCreator = ({ client, config }) => ({
         metadataType,
       )
       log.debug(`Getting missing ids for the following types: ${Object.keys(groupedElements)}`)
-      await Promise.all(
+      const results = await Promise.allSettled(
         Object.entries(groupedElements)
           .map(([typeName, typeElements]) => addMissingIds(client, typeName, typeElements))
       )
+      if (results.some(r => r.status === 'rejected')) {
+        throw new Error('Failed to add missing ids')
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/add_missing_ids.ts:87]: this means we won't get the original error anywhere.
we should either throw the original error as the code used to do before this change or log the error(s) first.

File: packages/salesforce-adapter/src/filters/add_missing_ids.ts [hunk 6]
```diff
@@ -84,10 +79,14 @@ const filter: FilterCreator = ({ client, config }) => ({
         metadataType,
       )
       log.debug(`Getting missing ids for the following types: ${Object.keys(groupedElements)}`)
-      await Promise.all(
+      const results = await Promise.allSettled(
         Object.entries(groupedElements)
           .map(([typeName, typeElements]) => addMissingIds(client, typeName, typeElements))
       )
+      const rejected = results.find(r => r.status === 'rejected') as PromiseRejectedResult | undefined
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/add_missing_ids.ts:86]: can't we use a type guard here?
using casts is not good practice

File: packages/salesforce-adapter/src/filters/add_missing_ids.ts [hunk 7]
```diff
@@ -84,10 +79,14 @@ const filter: FilterCreator = ({ client, config }) => ({
...
+      if (rejected !== undefined) {
+        throw new Error(rejected.reason)
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/add_missing_ids.ts:88]: I don't think this preserves the original stack, does it?

**Review by ori-moisis — COMMENTED** [review]
generally I think this is ok to merge
just need to make sure the target platform change doesn't break downstream projects.

I think we might need to mention the change in the target platform in the release notes

File: packages/salesforce-adapter/src/filters/add_missing_ids.ts (hunk 7 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/filters/add_missing_ids.ts]: v, rethrowing when the reason is a error

File: packages/salesforce-adapter/src/filters/add_missing_ids.ts (hunk 6 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/filters/add_missing_ids.ts]: v

File: packages/salesforce-adapter/src/filters/add_missing_ids.ts [hunk 8]
```diff
@@ -84,10 +85,14 @@ const filter: FilterCreator = ({ client, config }) => ({
         metadataType,
       )
       log.debug(`Getting missing ids for the following types: ${Object.keys(groupedElements)}`)
-      await Promise.all(
+      const results = await Promise.allSettled(
         Object.entries(groupedElements)
           .map(([typeName, typeElements]) => addMissingIds(client, typeName, typeElements))
       )
+      const rejected = results.find(isRejected)
+      if (rejected !== undefined && _.isError(rejected.reason)) {
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/add_missing_ids.ts:93]: do we need to take care of the case where `.reason` is not an error?

**Review by ori-moisis — APPROVED** [review]
code looks ok
please make sure everything builds correctly **before** merging

**Pull request closed without merging**